### ISSUE TITLE
Add support for dropping other tables in the DB on pull

### ISSUE THREAD

**Issue by bmcclure** [kalabox/kalabox#1755]
This is more a feature request than a bug I think.

I run into this scenario sometimes:
1. Make some site changes locally which delete one or more database tables (particularly Drupal fields)
2. Run kbox pull to pull down a version of the database which still has those fields in it
3. Try and delete those fields again
4. Error about field_deleted_data_* table already existing, and I have to manually drop the table and re-try the command for each one it finds.

Is there any way to either by default or via an option, to clear all existing tables out? Unless there's something kalabox needs that it won't get when it pulls the database down, perhaps just dropping and re-creating the database would be the simplest way to ensure there are no left over tables, though I'm not sure if maybe people have use cases with Kalabox for maintaining some tables locally that they wouldn't want to have lost when pulling.

**Comment by pirog** [comment]
@bmcclure im not really understanding the use case here. You can manually get the database locally if you want to edit it directly. Either with `kbox mysql` or `kbox services` and connecting with a third party tool like sqlpro or sql workbench.

There is also a skipdata option in the config you can use
http://pantheon.kalabox.io/en/v2.1/users/config/

**Comment by bmcclure** [comment]
The fact that I can use `kbox pull` to get a fresh database, but get get database errors from tables being left behind from the database that was there previously, seems like not behavior that would ever be desired.

My use case above is a really common one. I'll go into a little more detail:

1. I'm working on a site in Kalabox, and I delete some fields. When I do this, Drupal removes the field's table from the database, and creates a new table, prefixed with "field_deleted_data_". This is all stuff that Kalabox isn't, and shouldn't be, concerned with.
2. I realize I need some database changes from the client that are in Pantheon, so I end up needing to use `kbox pull` to pull a new copy of the files and database from Pantheon. This appears to work as expected, and I get the new database from Pantheon. The fields I deleted locally previously are back again, because they still exist in the database I pulled down.
3. Now I need to delete those fields again, but I can't (SQL error) because the field_deleted_data_* tables that Drupal created before I ran `kbox pull` still exist, even though they have nothing to do with the new database that Kalabox pulled down--they are in fact left over tables from the previous database.

I don't think this would ever be the desired behavior. If I pull down a fresh database from Pantheon, I expect my database to mirror Pantheon's, I don't want tables left behind that were created from Drupal locally but that aren't included in the database I'm pulling down, because it'll create an inconsistent database state and cause SQL errors like described above.

**Comment by pirog** [comment]
@bmcclure i see what you are saying. that said im not sure if we want to completely destroy the database every time we do a pull. i can **definitely** think of use cases where the behavior you are proposing would not be desirable and perhaps more to point workflows where you are relying on pushing databases around to deploy changes are almost always going to result in pain in one form or the other. 

We give access to the database so it should be pretty easy to manually go in and do what you need to or script something. It might be worth getting your DB changes up into the dev instance as soon as possible so you avoid this kind of mismatch.

**Comment by bmcclure** [comment]
@pirog To clarify, I do use an all-code approach for the most part. However, I still run into instances where I've started making a change that I plan to export to code, but realize there are changes other people have made upstream that I forgot to pull in first.

I don't ever push my database up from Kalabox to Pantheon, but I do pull the database down sometimes, either when I want to get the latest content/files, or when someone other than me has made config changes on the site. Typically then, I make all my config changes locally and export the config into code, commit, deploy, and import on Pantheon. However, once in a while I realize the database I'm making changes to isn't recent enough and before I export my config I pull a fresh database so I can start with a clean slate.

I know there may ultimately be better ways to handle some of this process, but we are trying to stick with a best-practices approach and still running into this issue at times even when the database _only_ ever flows downstream.

**Comment by bmcclure** [comment]
I consider my local Kalabox database ephemeral--I can pull down to overwrite it at anytime, and it never gets pushed back up. But this becomes problematic if it starts building up extra tables that aren't in the production DB. Someday if I need to pull the prod DB down to make a quick fix that I push back up, then suddenly the prod DB will have all of these extra dev tables.

**Comment by pirog** [comment]
@bmcclure for sure. i guess my points are:

1. Get your downstream (local) db changes into dev ASAP
2. Get other contributors to put their config in code (thus limiting the amount of times you might need to pull the DB down)
3. You can spin up multiple kalabox apps so its sometimes nice to have an instance where you can sandbox and new and big feature and another instance that is just the dev mainline (this is basically like a poor-mans-multidev)

I'm open to your change eg completely wiping the DB on every pull, id like to see more feedback here first

**Comment by bmcclure** [comment]
All 3 are good points, the third especially would be a great idea I think for me if I ever have to make a quick fix locally and push back up as there could be an environment dedicated to that sort of thing that won't have extra such tables laying around.

Today I ran into an issue that I don't know the best way to have gotten around (other than using your #3 suggestion): I exported my config to code, uploaded to Pantheon, but the changes were extensive and the server was running out of memory and causing the config import to error out. I pulled the database down so that I could verify whether I could import it locally or if there was some memory leak somewhere causing it not to be importable at all, but ran into this issue with all the fields I had deleted as part of that config update and it took me 30 minutes or so to chunk through one at a time when it errored out to drop the table and try again (because I wasn't sure what the names of the tables were it was going to error out on, it uses a hash value).

Anyway, long story short, I think following all 3 of your points closely would probably prevent most issues, but that having extra tables around in the local database still presents potential unexpected behavior and an inconsistent database state to the local Kalabox instance.

As one suggestion, what about an argument you could use with `kbox pull` such as `--clean` that would specify you want to make sure you have a completely clean version of the database you're pulling?

I'm definitely open to other ideas and hearing more feedback about something like this, and I'm not totally opposed to just scripting it in my shell or with Robo or Gulp if it's not something that would make sense to include in Kalabox for any reason.

**Comment by pirog** [comment]
Hey all!

We've (not so) secretly been working on a new project called [Lando](https://github.com/lando/lando) which is a completely different kind of local development tool but very much the spiritual successor to Kalabox. For that reason we are no longer actively fixing issues with Kalabox although we are still providing support via our [slack channel](https://slackpass.io/kalabox)

At this point we encourage you to make the switch to Lando, especially if you've posted an issue here in the Kalabox queue. There is a **VERY HIGH CHANCE** that this issue has been addressed in Lando and a **NEAR CERTAIN CHANCE** that it will not be back ported to Kalabox.

Please check out https://docs.lndo.io/ to get rolling with Lando and enjoy!
